### PR TITLE
Isolate TestFileSystemCache to fix its flakiness

### PR DESCRIPTION
**Pull request by jitheshtr**
## Description
`TrinoFileSystemCache` holds the global cache of filesystem objects within a running server. `TestFileSystemCache` expects that this singleton instance is available to itself and validates the number of entries in the cache accordingly. This is not a valid assumption as other threads running concurrently in the same JVM can add and delete filesystem objects in `TrinoFileSystemCache`. To avoid this interference between `TestFileSystemCache` and other threads running in the same JVM, we move the former to run in a separate runner instance in CI. This should avoid the flakiness reported in https://github.com/trinodb/trino/issues/17158

## Additional context and related issues
Fixes https://github.com/trinodb/trino/issues/17158

## Release notes

(X) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -482,6 +482,10 @@ jobs:
             !:trino-sqlserver,
             !:trino-test-jdbc-compatibility-old-server,
             !:trino-tests'
+      - name: Maven FileSystemCache Tests
+        # Isolate HDFS file system cache concurrency test to avoid flakiness
+        run: |
+          $MAVEN test ${MAVEN_TEST} -pl :trino-hdfs -P test-hdfs-cache
```
Comment (findepi) [inline .github/workflows/ci.yml:488]: regardless of GIB?

what if test-hdfs-cache is removed from "other modules"?

File: lib/trino-hdfs/pom.xml [hunk 2]
```diff
@@ -225,4 +225,42 @@
...
+        <profile>
+            <id>test-hdfs-cache</id>
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-surefire-plugin</artifactId>
+                        <configuration>
+                            <includes>
+                                <include>**/TestFileSystemCache.java</include>
```
Comment (findepi) [inline lib/trino-hdfs/pom.xml:258]: Why is separate profile needed? do you maybe understand why https://github.com/trinodb/trino/pull/17266 was not sufficient?

Comment (jitheshtr) [reply lib/trino-hdfs/pom.xml:258]: One problem i see with #17266 is that the thread count is limited only for trino-hdfs module. There is a possibility that the tests in `trino-hdfs` and `trino-hive-hadoop2` can run parallelly in separate threads in the same JVM interacting with same global `TrinoFileSystemCache`. This can cause `testFileSystemCacheConcurrency()` to fail.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (jitheshtr) [reply .github/workflows/ci.yml:488]: Not sufficiently versed with GIB, need to get back on this.

File: lib/trino-hdfs/pom.xml (hunk 2 above)
Comment (findepi) [reply lib/trino-hdfs/pom.xml:258]: > There is a possibility that the tests in `trino-hdfs` and `trino-hive-hadoop2` can run parallelly in separate threads in the same JVM interacting with same global `TrinoFileSystemCache`.

i'd hope different modules do not share same JVM for test execution.
I think maven surefire plugin starts another JVM for test run.

Comment (jitheshtr) [reply lib/trino-hdfs/pom.xml:258]: I was able to spot an interaction from another thread that invoked `getFileSystem()` while the tests from `TestFileSystemCache` were running (From https://github.com/jitheshtr/trino/actions/runs/4964921803/jobs/8885392660?pr=6) -
```
2023-05-12T21:50:19.774-0600 SEVERE Could not process download requests
java.lang.RuntimeException: Saw user runner with null proxy user
	at io.trino.hdfs.TrinoFileSystemCache.createFileSystemKey(TrinoFileSystemCache.java:259)
	at io.trino.hdfs.TrinoFileSystemCache.getInternal(TrinoFileSystemCache.java:115)
	at io.trino.hdfs.TrinoFileSystemCache.get(TrinoFileSystemCache.java:94)
	at org.apache.hadoop.fs.ForwardingFileSystemCache.get(ForwardingFileSystemCache.java:38)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:479)
	at com.qubole.rubix.bookkeeper.FileDownloader.getFileDownloadRequestChains(FileDownloader.java:109)
	at com.qubole.rubix.bookkeeper.RemoteFetchProcessor.processRequest(RemoteFetchProcessor.java:155)
	at com.qubole.rubix.bookkeeper.RemoteFetchProcessor.runOneIteration(RemoteFetchProcessor.java:144)
	at com.google.shaded.shaded.common.util.concurrent.AbstractScheduledService$ServiceDelegate$Task.run(AbstractScheduledService.java:193)
	at com.google.shaded.shaded.common.util.concurrent.Callables$4.run(Callables.java:119)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

This exception was triggered by the code changes here - https://github.com/jitheshtr/trino/pull/6/commits/a1042a50953c512248e64696e00c11a266ca259a

Comment (jitheshtr) [reply lib/trino-hdfs/pom.xml:258]: >i'd hope different modules do not share same JVM for test execution.
I think maven surefire plugin starts another JVM for test run.

You are right, from https://maven.apache.org/surefire/maven-surefire-plugin/examples/fork-options-and-parallel-execution.html it should invoke different modules in different JVMs -
> The default setting is forkCount=1/reuseForks=true, which means that maven-surefire-plugin creates one new JVM process to execute all tests in one Maven module.

Looking at above capture, simultaneous use of `TrinoFileSystemCache` from a different thread is happening from Rubix code running in the same JVM.

Comment (findepi) [reply lib/trino-hdfs/pom.xml:258]: let's capture the motivation (ie something mentioning "rubix") in the commit message AND a code comment in the profile definition inside the pom file

**Review by xkrogen — COMMENTED** [review]
It's a shame that Maven doesn't have the option to isolate certain tests/suites to run in separate JVMs -- the Spark project leverages `sbt`'s ability to do this and it's quite nice ([check out an example here](https://github.com/apache/spark/pull/29059/files)).

I am wondering if we can at least introduce a more generic/scalable mechanism than what is shown here ... It seems a bit awkward to have a whole GH Action step for a single test class.

What if we did this:
- Create a property like `isolatedJvmTests` which, for now, contains just `TestFileSystemCache`
- Use this property as an exclude for the `default` profile surefire plugin
- Rename the new profile to `test-isolated-jvm-suites` and leverage the property there for the include list
- Use [Surefire's fork options](https://maven.apache.org/surefire/maven-surefire-plugin/examples/fork-options-and-parallel-execution.html#forked-test-execution) with `reuseForks=false` and `forkCount=1`. This forces Surefire to build a new JVM for each test suite

For just `TestFileSystemCache` we have the same end result, but now we have a single place we can easily update to add any new suites in the future which have a similar need.

WDYT? Feel free to tell me it's overkill :)

**Comment by jitheshtr** [comment]
> It's a shame that Maven doesn't have the option to isolate certain tests/suites to run in separate JVMs -- the Spark project leverages `sbt`'s ability to do this and it's quite nice ([check out an example here](https://github.com/apache/spark/pull/29059/files)).
> 
> I am wondering if we can at least introduce a more generic/scalable mechanism than what is shown here ... It seems a bit awkward to have a whole GH Action step for a single test class.
> 
> What if we did this:
> 
>     * Create a property like `isolatedJvmTests` which, for now, contains just `TestFileSystemCache`
> 
>     * Use this property as an exclude for the `default` profile surefire plugin
> 
>     * Rename the new profile to `test-isolated-jvm-suites` and leverage the property there for the include list
> 
>     * Use [Surefire's fork options](https://maven.apache.org/surefire/maven-surefire-plugin/examples/fork-options-and-parallel-execution.html#forked-test-execution) with `reuseForks=false` and `forkCount=1`. This forces Surefire to build a new JVM for each test suite
> 
> 
> For just `TestFileSystemCache` we have the same end result, but now we have a single place we can easily update to add any new suites in the future which have a similar need.
> 
> WDYT? Feel free to tell me it's overkill :)

Looking at the surefire plugins inclusion/exclusion [documentation](https://maven.apache.org/surefire/maven-surefire-plugin/examples/inclusion-exclusion.html) - all of them are file/class name based exclusion. Could you please help identify an example of property based exclusion for maven-surefire-plugin?

**Review by findepi — APPROVED** [review]

**Comment by hashhar** [comment]
> Looking at the surefire plugins inclusion/exclusion [documentation](https://maven.apache.org/surefire/maven-surefire-plugin/examples/inclusion-exclusion.html) - all of them are file/class name based exclusion. Could you please help identify an example of property based exclusion for maven-surefire-plugin?

I think he means to have something like:

```xml
<properties>
  <isolatedJvmTests>**/TestFileSystemCache.java,**/SomeOther.java</isolatedJvmTests>
</properties>
...
<profiles>
  <profile>
    <id>default</id>
    <build>
      <plugins>
        <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-surefire-plugin</artifactId>
          <configuration>
            <excludes>
              <exclude>${isolatedJvmTests}</exclude>
            </excludes>
          </configuration>
        </plugin>
      </plugins>
    </build>
  </profile>

  <profile>
    <id>test-isolated-jvm-suites</id>
    <build>
      <plugins>
        <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-surefire-plugin</artifactId>
          <configuration>
            <includes>
              <include>${isolatedJvmTests}</include>
            </includes>
            <reuseForks>false</reuseForks>
            <forkCount>1</forkCount>
          </configuration>
        </plugin>
      </plugins>
    </build>
  </profile>
</profiles>
```

**Comment by phd3** [comment]
IMO #17266 should've been sufficient, but looking at the stacktraces, imo the issue could be some missing cleanup in Rubix tests. So even though tests are single threaded, some unclosed executor from rubix is bothering this test later, and surfaced  here as a flake. I tried to look around that code - but can't put my finger on what's we are failing to shutting down properly. 
 But it seems reasonable to completely isolate this one, since it depends on global state.

File: lib/trino-hdfs/pom.xml [hunk 3]
```diff
@@ -16,6 +16,9 @@
         <air.main.basedir>${project.parent.basedir}</air.main.basedir>
         <!-- e.g. TestFileSystemCache asserts on static global state -->
         <air.test.thread-count>1</air.test.thread-count>
```
Comment (phd3) [inline lib/trino-hdfs/pom.xml:18]: we can revert this change as well

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -482,6 +482,10 @@ jobs:
             !:trino-sqlserver,
             !:trino-test-jdbc-compatibility-old-server,
             !:trino-tests'
+      - name: Maven FileSystemCache Tests
```
Comment (phd3) [inline .github/workflows/ci.yml:485]: nit: May be just call this `trino-hdfs isolated JVM tests`

**Review by phd3 — APPROVED** [review]
Thanks @jitheshtr

**Comment by phd3** [comment]
Failures seem unrelated : https://github.com/trinodb/trino/issues/16277 , https://github.com/trinodb/trino/issues/17523

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -482,6 +482,10 @@ jobs:
             !:trino-sqlserver,
             !:trino-test-jdbc-compatibility-old-server,
             !:trino-tests'
+      - name: trino-hdfs isolated JVM tests
+        # Isolate HDFS file system cache concurrency test to avoid flakiness
+        run: |
+          $MAVEN test ${MAVEN_TEST} -pl :trino-hdfs -P test-isolated-jvm-suites
```
Comment (phd3) [inline .github/workflows/ci.yml:488]: @hashhar @nineinchnick This and previous step will both write to `trino-hdfs/target/surefire-reports/`. However I do see different `TEST-TestSuite.xml` and `TEST-io.trino.hdfs.TestFileSystemCache.xml` that are relevant for "Upload test report".  Also "Upload test results" is only invoked on failures - so we'll upload the contents for the first profile that failed.  From perspective of CI's further dependencies - is a separate job preferred or this is fine?

Comment (hashhar) [reply .github/workflows/ci.yml:488]: @nineinchnick can you verify. As i understand we'll have both reports and if first job fails only that report will be uploaded but if the latter fails then both first and second will be uploaded.

I don't think it's a problem though since the worst that can happen is you'll have a report for a successful run as well which doesn't hurt.

Comment (nineinchnick) [reply .github/workflows/ci.yml:488]: Yes, I think it's fine as is.

Comment (phd3) [reply .github/workflows/ci.yml:488]: Thanks for the input
